### PR TITLE
Use power of two for underlying storage of CircularBuffer

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We should use power of two for the underlying storage to allow us to make use of bitmasking which is faster then modulo operations.

Modifications:

Use power of two and replace module with bitmasking.

Result:

Faster implementation of CircularBuffer.

### CODE REVIEW

File: Sources/NIO/CircularBuffer.swift [hunk 1]
```diff
@@ -20,6 +20,28 @@ public protocol AppendableCollection: Collection {
...
+
+        var n = self
+
+        n -= 1
+        n |= n >> 1
+        n |= n >> 2
+        n |= n >> 4
+        n |= n >> 8
+        n |= n >> 16
+        n |= n >> 32
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: yeah thats a good point... Let me just use UInt32 and move on. That should be plenty of space. WDYT ?

**Comment by normanmaurer** [comment]
@weissi I see no differences with our network benchmarks (which is not really unexpected atm). But I think we all know that modulo is considered slow compared to bit masking.

File: Sources/NIO/CircularBuffer.swift [hunk 2]
```diff
@@ -41,8 +63,9 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
     /// Allocates a buffer that can hold up to `initialRingCapacity` elements and initialise an empty ring backed by
     /// the buffer. When the ring grows to more than `initialRingCapacity` elements the buffer will be expanded.
     public init(initialRingCapacity: Int) {
-        self.bufferCapacity = initialRingCapacity
-        self.buffer = ContiguousArray<E?>(repeating: nil, count: Int(initialRingCapacity))
+        let capacity = initialRingCapacity.nextPowerOf2()
```
Comment (weissi) [inline Sources/NIO/CircularBuffer.swift]: do the tests really pass with this? The need a careful look if the pass as they definitely try to test the edge cases on the edge of extension

Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: yep they pass

**Comment by weissi** [comment]
@normanmaurer sure, I thought a micro benchmark but it's unlikely to be worse so maybe it's fine

File: Sources/NIO/CircularBuffer.swift (hunk 2 above)
Comment (weissi) [reply Sources/NIO/CircularBuffer.swift]: @normanmaurer then they will most likely need adjustment to still test the edge cases (unless they happen to have picked only powers of 2). Did you check the code coverage?

Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: nope not yet

**Comment by normanmaurer** [comment]
@Lukasa @weissi PTAL again... I also did a few more changes to make the whole data structure a bit smaller.

File: Sources/NIO/CircularBuffer.swift [hunk 3]
```diff
@@ -129,17 +123,17 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
...
     /// Returns whether the ring is empty.
     public var isEmpty: Bool {
-        return self.ringLength == 0
+        return self.startIdx == self.endIdx
     }
 
     /// Returns the number of element in the ring.
     public var count: Int {
-        return self.ringLength
+        return (self.endIdx - self.startIdx) & self.mask
```
[the remark this answers is not in the document]
Comment (weissi) [reply Sources/NIO/CircularBuffer.swift]: yes, that's correct. Can we also have a test case for that please

[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: I think it is actually correct ( the impl) let me add a unit test for it and either prove me or you wrong

[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: ok added test for it... works as expected (thanks to bitwise ops ;))

**Comment by normanmaurer** [comment]
Will add more tests tomorrow

File: Sources/NIO/CircularBuffer.swift [hunk 4]
```diff
@@ -26,62 +26,59 @@ public protocol AppendableCollection: Collection {
...
-        let expandBuf: Bool = self.bufferCapacity == self.ringLength
+        self.buffer[self.endIdx] = value
+        self.endIdx = (self.endIdx + 1) & self.mask
 
-        if expandBuf {
+        if startIdx == endIdx {
+            // No more room left for another append so grow the buffer now.
             var newBacking: ContiguousArray<E?> = []
-            let newCapacity = Swift.max(1, 2 * self.bufferCapacity)
+            let newCapacity = self.buffer.count << 1 // Double the storage.
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: @weissi IMHO its not really guaranteed its as fast... generally speaking bitshifting should be faster.

[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIO/CircularBuffer.swift]: Any optimising compiler in the land will trivially see that multiply-by-two is the same as bit shift left by one.

File: Tests/NIOTests/CircularBufferTests.swift [hunk 5]
```diff
@@ -207,4 +207,70 @@ class CircularBufferTests: XCTestCase {
...
+
+
+
+
+
+
+
+
+
+
```
Comment (Lukasa) [inline Tests/NIOTests/CircularBufferTests.swift]: Um?

File: Sources/NIO/CircularBuffer.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: @Lukasa I learned the hard way to not trust these ;) That said if you are both strong I can remove it... that said I think its not really hard to understand as it is.

File: Sources/NIO/CircularBuffer.swift [hunk 6]
```diff
@@ -26,62 +26,60 @@ public protocol AppendableCollection: Collection {
 public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
     private var buffer: ContiguousArray<E?>
 
-    /// The capacity of the underlying buffer
-    private var bufferCapacity: Int
-
     /// The index into the buffer of the first item
-    private var startIdx = 0
+    private(set) /* private but tests */ internal var startIdx = 0
```
Comment (Lukasa) [inline Sources/NIO/CircularBuffer.swift]: If you're going to expose these as internal I think these variables need different names. They are dangerously close to shadowing the variables `startIndex` and `endIndex` that come from `Collection`, and yet they behave quite differently.

File: Sources/NIO/CircularBuffer.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIO/CircularBuffer.swift]: 👍

File: Sources/NIO/CircularBuffer.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIO/CircularBuffer.swift]: I mean, this is a trivial peephole optimisation. :wink:

File: Tests/NIOTests/CircularBufferTests.swift (hunk 5 above)
Comment (normanmaurer) [reply Tests/NIOTests/CircularBufferTests.swift]: more empty lines ftw... seriously let me cleanup while adding more tests

File: Sources/NIO/CircularBuffer.swift (hunk 6 above)
Comment (Lukasa) [reply Sources/NIO/CircularBuffer.swift]: Up to @weissi.

File: Tests/NIOTests/CircularBufferTests.swift (hunk 5 above)
Comment (normanmaurer) [reply Tests/NIOTests/CircularBufferTests.swift]: I wonder if I felt asleep while holding the enter key 🤷‍♂️

File: Sources/NIO/CircularBuffer.swift (hunk 6 above)
Comment (weissi) [reply Sources/NIO/CircularBuffer.swift]: I like `headIdx` and `tailIdx`, then it's also more obvious that head can be larger than tail

**Comment by normanmaurer** [comment]
@weissi @Lukasa I enhanced the tests a bit and addressed comments. Can you let me know what kind of other tests you would like to see (if any)?

File: Sources/NIO/CircularBuffer.swift [hunk 7]
```diff
@@ -26,62 +26,59 @@ public protocol AppendableCollection: Collection {
...
-    private var startIdx = 0
+    private(set) /* private but tests */ internal var headIdx = 0
 
     /// The index into the buffer of the next free slot
-    private var endIdx = 0
+    private(set) /* private but tests */ internal var tailIdx = 0
 
-    /// The number of items in the ring part of this buffer
-    private var ringLength = 0
+    private var mask: Int {
```
Comment (Lukasa) [inline Sources/NIO/CircularBuffer.swift]: Can you add a doc comment for `mask`?

Comment (normanmaurer) [reply Sources/NIO/CircularBuffer.swift]: addressed

**Comment by normanmaurer** [comment]
@weissi @Lukasa PTAL again
